examples/cis3-nft-sponsored-txs/src/lib.rs

```diff
@@ -0,0 +1,2228 @@
+    /// transactions.
+    nonces:              StateMap<AccountAddress, u64, S>,
+}
+
+/// Part of the parameter type for the contract function `registerPublicKey`.
+/// Takes the account and the public key that should be linked.
+#[derive(Debug, Serialize, SchemaType)]
+pub struct RegisterPublicKeyParam {
+    /// Account that a public key will be registered to.
+    account:    AccountAddress,
+    /// The public key that should be linked to the above account.
+    public_key: PublicKeyEd25519,
+}
+
+/// The parameter type for the contract function `setImplementors`.
+/// Takes a standard identifier and list of contract addresses providing
+/// implementations of this standard.
+#[derive(Debug, Serialize, SchemaType)]
+struct SetImplementorsParams {
+    /// The identifier for the standard.
+    id:           StandardIdentifierOwned,
+    /// The addresses of the implementors of the standard.
+    implementors: Vec<ContractAddress>,
+}
+
+/// The parameter type for the contract function `permitTransfer`.
+/// Takes a signature, all input parameters to the regular `transfer` function,
+/// the `contract_address`/`entry_point` that the signature is intended for, and
+/// a `nonce` to prevent replay attacks.
+#[derive(SchemaType, Serialize, Debug)]
+pub struct PermitTransferParam {
+    /// A signature generated from the corresponding public key that is
+    /// registered in this contract.
+    signature:        SignatureEd25519,
+    /// The `from` parameter from the regular `transfer` function. Because only
+    /// an `accountAddress` can generate a signature, the `from` has
+    /// to be an `accountAddress`.
+    from:             AccountAddress,
+    /// The `to` parameter from the regular `transfer` function.
+    to:               Receiver,
+    /// The `data` parameter from the regular `transfer` function.
+    data:             AdditionalData,
+    /// The `amount` parameter from the regular `transfer` function.
+    amount:           ContractTokenAmount,
+    /// The `token_id` parameter from the regular `transfer` function.
+    token_id:         ContractTokenId,
+    /// The contract_address that the signature is intended for.
+    contract_address: ContractAddress,
+    /// The entry_point that the signature is intended for.
+    entry_point:      OwnedEntrypointName,
+    /// A nonce to prevent replay attacks.
+    nonce:            u64,
+}
+
+/// The message struct that is signed in the `permitTransfer` function.
+/// Takes all input parameters to the regular `transfer` function,
+/// the `contract_address`/`entry_point` that the signature is intended for, and
+/// a `nonce` to prevent replay attacks.
+#[derive(SchemaType, Serialize)]
+struct PermitTransferMessage {
+    /// The `from` parameter from the regular `transfer` function. Because only
+    /// an `accountAddress` can generate a signature, the `from` has
+    /// to be an `accountAddress`.
+    from:             AccountAddress,
```

Comment:
Support transfers from contracts would mean that the smart contract needs to `sign` a transaction but we don't have a standard yet for that. For now, it seems that the `permit` function can be used by only `AccountAddresses` since they have public keys and can sign.